channel2/wss_listener.go

```diff
@@ -0,0 +1,280 @@
+/*
+	Copyright NetFoundry, Inc.
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+	https://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+package channel2
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"github.com/gorilla/mux"
+	"github.com/gorilla/websocket"
+	"github.com/michaelquigley/pfxlog"
+	"github.com/openziti/foundation/identity/identity"
+	"github.com/openziti/foundation/transport"
+	"io"
+	"net"
+	"net/http"
+	"time"
+)
+
+type wssListener struct {
+	identity       *identity.TokenId
+	endpoint       transport.Address
+	socket         io.Closer
+	close          chan struct{}
+	handlers       []ConnectionHandler
+	created        chan Underlay
+	connectOptions ConnectOptions
+	serverCert     string
+	key            string
+	tcfg           transport.Configuration
+	incoming       chan *WssConnection
+}
+
+func NewWssListener(identity *identity.TokenId, endpoint transport.Address, connectOptions ConnectOptions, serverCert string, key string) UnderlayListener {
+	return NewWssListenerWithTransportConfiguration(identity, endpoint, connectOptions, nil, serverCert, key)
+}
+
+func NewWssListenerWithTransportConfiguration(identity *identity.TokenId, endpoint transport.Address, connectOptions ConnectOptions, tcfg transport.Configuration, serverCert string, key string) UnderlayListener {
+	return &wssListener{
+		identity:       identity,
+		endpoint:       endpoint,
+		close:          make(chan struct{}),
+		created:        make(chan Underlay),
+		incoming:       make(chan *WssConnection, connectOptions.MaxQueuedConnects),
+		connectOptions: connectOptions,
```

Comment:
These `connectOptions` are meant to deal with TCP/TLS connection limits. `http.Server` has its own set of tunable settings. I suspect we don't need `connectOptions` here.